### PR TITLE
hubble: Add support for filtering on HTTP headers

### LINKED ISSUE #776: Support for filtering on HTTP headers
HTTP flows [contain headers](https://github.com/cilium/cilium/blob/862e191b3574e4a7c5d2046d28d980c163b50a6f/api/v1/flow/flow.proto#L456-L469) but Hubble doesn't support filtering flows based on HTTP headers. Using the CLI, we can already filter based on HTTP status code, methods and paths but filtering on headers is still missing.

### PR DESCRIPTION
**Pull request by rohithsai1904**
Please ensure your pull request adheres to the following guidelines:

- [ ] For first time contributors, read [Submitting a pull request](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#submitting-a-pull-request)
- [ ] All code is covered by unit and/or runtime tests where feasible.
- [ ] All commits contain a well written commit description including a title,
      description and a `Fixes: #XXX` line if the commit addresses a particular
      GitHub issue.
- [ ] If your commit description contains a `Fixes: <commit-id>` tag, then
      please add the commit author[s] as reviewer[s] to this issue.
- [ ] All commits are signed off. See the section [Developer’s Certificate of Origin](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#dev-coo)
- [ ] Provide a title or release-note blurb suitable for the release notes.
- [ ] Thanks for contributing!

Fixes: cilium/hubble#776 

Added filter for HTTP support
Need to add flags in hubble cli
```release-note
hubble: Add Support for filtering on HTTP headers
```

### CODE REVIEW

File: api/v1/flow/flow.proto [hunk 1]
```diff
@@ -479,6 +481,9 @@ message FlowFilter {
 
     // trace_id filters flows by trace ID
     repeated string trace_id = 28;
+
+    //http_header filters flows based on HTTPheader
+    repeated HTTPHeader http_header=29;
```
Comment (kaworu) [inline api/v1/flow/flow.proto]: nit:
```suggestion
    // http_header filters flows based on a list of HTTP Headers
    repeated HTTPHeader http_header = 29;
```

File: api/v1/flow/flow.pb.go [hunk 2]
```diff
@@ -2781,7 +2783,7 @@ func (x *FlowFilter) GetDestinationIp() []string {
 	return nil
 }
 
-func (x *FlowFilter) GetDestinationPod() []string {
+func (x *FlowFilter) GetDestinationPod() []string
```
Comment (kaworu) [inline api/v1/flow/flow.pb.go]: this leads to a compilation error.

File: pkg/hubble/filters/httpheader.go [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+	}
+
+}
+
+// HTTPHeaderimplements filtering based on endpoint workload
+
+type HTTPHeader struct{}
+
+func (*HTTPHeader) OnBuildFilter(ctx context.Context, ff *flowpb.FlowFilter) ([]FilterFunc, error)
+{
```
Comment (kaworu) [inline pkg/hubble/filters/httpheader.go]: The `{` needs to be on the same line here

```suggestion
func (*HTTPHeader) OnBuildFilter(ctx context.Context, ff *flowpb.FlowFilter) ([]FilterFunc, error) {
```

File: pkg/hubble/filters/httpheader.go [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+
+	for _, t := range types {
+		if t.GetType() == api.MessageTypeAccessLog {
+			return true
+		}
+	}
+
+	return false
+}
+
```
Comment (kaworu) [inline pkg/hubble/filters/httpheader.go]: `httpMatchCompatibleEventFilter` is already defined in this package (see [here](https://github.com/cilium/cilium/blob/bf3532efd47f45d505f9f1dffdda1e3b3df13662/pkg/hubble/filters/http.go#L19)).


```suggestion
```

File: pkg/hubble/filters/httpheader.go [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+		for _, w := range gethttp(ev).GetHeaders() {
+			for _, f := range h {
+				if (f.GetKey() == "" || f.GetKey() == w.GetKey()) &&
+					(f.GetValue() == "" || f.GetValue() == w.GetValue()) {
+					return true
+				}
+			}
+		}
+		return false
+	}
```
Comment (kaworu) [inline pkg/hubble/filters/httpheader.go]: the returned values are `(FilterFunc, error)`

```suggestion
	}, nil
```

File: pkg/hubble/filters/httpheader.go [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+		if !httpMatchCompatibleEventFilter(ff.GetEventType()) {
+			return nil, errors.New("filtering by http headers requires " +
+				"the event type filter to only match 'l7' events")
+		}
+		hh, err := filterByHTTPHeader(ff.GetHttpHeader())
+		if err != nil {
+			return nil, fmt.Errorf("invalid http header filter: %v", err)
+		}
+		fs = append(fs, hh)
+	}
```
Comment (kaworu) [inline pkg/hubble/filters/httpheader.go:47]:
```suggestion
	}
	return fs, nil
```

File: pkg/hubble/filters/httpheader.go [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+		}
+	}
+
+	return false
+}
+
+func filterByHTTPHeader(h []*flowpb.HTTPHeader,gethttp func(*v1.Event) *flowpb.HTTP) (FilterFunc, error) {
+
+	return func(ev *v1.Event) bool {
+		for _, w := range gethttp(ev).GetHeaders() {
```
Comment (kaworu) [inline pkg/hubble/filters/httpheader.go]: `filterByHTTPHeader` is called without the `gethttp` `func` param, and I think we could do without, e.g.

```suggestion
		for _, w := range ev.GetFlow().L7.GetHttp().GetHeaders() {
```

File: pkg/hubble/filters/httpheader.go [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+			}
+		}
+		return false
+	}
+
+}
+
+// HTTPHeaderimplements filtering based on endpoint workload
+
+type HTTPHeader struct{}
```
Comment (kaworu) [inline pkg/hubble/filters/httpheader.go]: this type name must match the one used in `pkg/hubble/filters/filters.go`

```suggestion
type HTTPHeaderFilter struct{}
```

**Review by kaworu — CHANGES_REQUESTED** [review]
Thanks for the PR @rohithsai1904! I think the best way forward now is to add unit testing for the HTTP header filter once all the compiling issues are sorted out.

You could take inspiration from https://github.com/cilium/cilium/blob/bf3532efd47f45d505f9f1dffdda1e3b3df13662/pkg/hubble/filters/workload_test.go as it also test a key/value type of filter, although the http headers case is slightly simpler as it only appear once (vs workload filters which are doubled for source and destination).

**Comment by kaworu** [comment]
> Need to add unit test for http Header Filter

Thanks @rohithsai1904! Marking this PR as draft until it's ready for review with the unit tests.

**Comment by rohithsai1904** [comment]
@kaworu can you give me some specific testcases for the http header?

**Comment by kaworu** [comment]
@rohithsai1904 sure, as I mentionned above, structurally

> You could take inspiration from https://github.com/cilium/cilium/blob/bf3532efd47f45d505f9f1dffdda1e3b3df13662/pkg/hubble/filters/workload_test.go as it also test a key/value type of filter, although the http headers case is slightly simpler as it only appear once (vs workload filters which are doubled for source and destination).

As for HTTP headers examples, here are some:
```
Content-Type: text/html
Content-Length: 162
Connection: keep-alive
```

Also an interesting case specific to HTTP headers is that it's possible to have two of them with the same name, e.g.

```
Cache-control: no-cache
Cache-control: no-store
```

**Comment by rohithsai1904** [comment]
image

Stuck with this error

**Comment by rohithsai1904** [comment]
the function httpFlow is like
 
image

**Comment by rohithsai1904** [comment]
@kaworu @rolinh Can you please check the PR?

File: api/v1/flow/flow.proto [hunk 9]
```diff
@@ -481,6 +483,9 @@ message FlowFilter {
 
     // trace_id filters flows by trace ID
     repeated string trace_id = 28;
+
+    //http_header filters flows based on list of HTTPHeaders
+    repeated HTTPHeader http_header=29;
```
Comment (kaworu) [inline api/v1/flow/flow.proto]: Probably make sense to move this field with the other http related ones

```suggestion
    // http_header filters flows based on HTTP Headers
    repeated HTTPHeader http_header = 29;
```

File: pkg/hubble/filters/httpheader.go [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+	"context"
+	"errors"
+	"fmt"
+
+	flowpb "github.com/cilium/cilium/api/v1/flow"
+	v1 "github.com/cilium/cilium/pkg/hubble/api/v1"
+)
+
+func filterByHTTPHeader(h []*flowpb.HTTPHeader) (FilterFunc, error) {
+
```
Comment (kaworu) [inline pkg/hubble/filters/httpheader.go]:
```suggestion
```

File: pkg/hubble/filters/httpheader.go [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+	return func(ev *v1.Event) bool {
+		for _, w := range ev.GetFlow().L7.GetHttp().GetHeaders() {
```
Comment (kaworu) [inline pkg/hubble/filters/httpheader.go:17]: sorry my mistake from a previous suggestion
```suggestion
		for _, w := range ev.GetFlow().GetL7().GetHttp().GetHeaders() {
```

File: pkg/hubble/filters/httpheader_test.go [hunk 12]
```diff
@@ -0,0 +1,303 @@
...
+	}{
+		//header filter
+		{
+			name: "header-nil",
+			args: args{
+				f: []*flowpb.FlowFilter{
+					{
+						HttpHeader: []*flowpb.HTTPHeader{
+							{
+								Key:   "Content_Length",
```
Comment (kaworu) [inline pkg/hubble/filters/httpheader_test.go]:
```suggestion
								Key:   "Content Length",
```

ditto for the other test cases

File: pkg/hubble/filters/httpheader_test.go [hunk 13]
```diff
@@ -0,0 +1,303 @@
...
+
+			want: false,
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			fl, err := BuildFilterList(context.Background(), tt.args.f, []OnBuildFilter{&HTTPHeaderFilter{}})
+			assert.NoError(t, err)
+			assert.Equal(t, tt.want, fl.MatchOne(tt.args.ev))
```
Comment (kaworu) [inline pkg/hubble/filters/httpheader_test.go:333]: The compilation error is here: [`fl.MatchOne` expect a `*v1.Event` parameter](https://github.com/cilium/cilium/blob/76837ead2978467e1033ee40c35d7a26beadf39d/pkg/hubble/filters/filters.go#L38) but `tt.args.ev` is a slice.

**Review by kaworu — CHANGES_REQUESTED** [review]
Hi @rohithsai1904, I left some review comments. Please also avoid introducing blank lines and the Travis-CI error about go code formatting is [here](https://app.travis-ci.com/github/cilium/cilium/jobs/591328015#L407).

**Comment by rohithsai1904** [comment]
Hey @kaworu Can you please check with whats wrong? It could be great if u help me to merge and close issue this weekend.

File: api/v1/flow/flow.proto [hunk 14]
```diff
@@ -443,6 +443,7 @@ message FlowFilter {
     repeated Verdict verdict = 5;
     // event_type is the list of event types to filter on
     repeated EventTypeFilter event_type = 6;
+
```
Comment (kaworu) [inline api/v1/flow/flow.proto:446]:
```suggestion
```

File: api/v1/flow/flow.proto [hunk 15]
```diff
@@ -466,9 +467,13 @@ message FlowFilter {
     // GET, POST, PUT, etc. methods. This type of field is well suited for an
     // enum but every single existing place is using a string already.
     repeated string http_method = 21;
+
```
Comment (kaworu) [inline api/v1/flow/flow.proto:470]:
```suggestion
```

File: api/v1/flow/flow.proto [hunk 16]
```diff
@@ -466,9 +467,13 @@ message FlowFilter {
     // GET, POST, PUT, etc. methods. This type of field is well suited for an
     // enum but every single existing place is using a string already.
     repeated string http_method = 21;
+    
     // http_path is a list of regular expressions to filter on the HTTP path.
     repeated string http_path = 22;
```
Comment (kaworu) [inline api/v1/flow/flow.proto:473]:
```suggestion
```

File: api/v1/flow/flow.proto [hunk 17]
```diff
@@ -481,6 +486,8 @@ message FlowFilter {
 
     // trace_id filters flows by trace ID
     repeated string trace_id = 28;
+
+
```
Comment (kaworu) [inline api/v1/flow/flow.proto:490]:
```suggestion
```

File: pkg/hubble/filters/httpheader_test.go [hunk 18]
```diff
@@ -0,0 +1,337 @@
...
+						},
+					},
+				},
+			},
+
+			want: true,
+		},
+
+		{
+			name: "header_key_and_value_empty_match",
```
Comment (kaworu) [inline pkg/hubble/filters/httpheader_test.go:76]:
```suggestion
			name: "header-key-and-value-empty-match",
```

File: pkg/hubble/filters/httpheader_test.go [hunk 19]
```diff
@@ -0,0 +1,337 @@
...
+	v1 "github.com/cilium/cilium/pkg/hubble/api/v1"
+	"github.com/cilium/cilium/pkg/monitor/api"
+	"github.com/stretchr/testify/assert"
+)
+
+func TestHTTPHeaderFilter(t *testing.T) {
+
+	type args struct {
+		f  []*flowpb.FlowFilter
+		ev []*v1.Event
```
Comment (kaworu) [inline pkg/hubble/filters/httpheader_test.go:17]: this is causing a compilation failure, as the declared args `ev` are `*v1.Event` now
```suggestion
		ev *v1.Event
```

File: pkg/hubble/filters/httpheader_test.go [hunk 20]
```diff
@@ -0,0 +1,337 @@
...
+						},
+					},
+				},
+			},
+
+			want: true,
+		},
+
+		{
+			name: "header_key_match",
```
Comment (kaworu) [inline pkg/hubble/filters/httpheader_test.go:113]:
```suggestion
			name: "header-key-match",
```

File: pkg/hubble/filters/httpheader_test.go [hunk 21]
```diff
@@ -0,0 +1,337 @@
...
+						},
+					},
+				},
+			},
+
+			want: true,
+		},
+
+		{
+			name: "header_value_match",
```
Comment (kaworu) [inline pkg/hubble/filters/httpheader_test.go:148]:
```suggestion
			name: "header-value-match",
```

File: pkg/hubble/filters/httpheader_test.go [hunk 22]
```diff
@@ -0,0 +1,337 @@
...
+							},
+						},
+					},
+				},
+			},
+
+			want: true,
+		},
+		{
+			name: "header_key_and_value_match",
```
Comment (kaworu) [inline pkg/hubble/filters/httpheader_test.go:182]:
```suggestion
			name: "header-key-and-value-match",
```

File: pkg/hubble/filters/httpheader_test.go [hunk 23]
```diff
@@ -0,0 +1,337 @@
...
+						},
+					},
+				},
+			},
+
+			want: true,
+		},
+
+		{
+			name: "header_key_mismatch",
```
Comment (kaworu) [inline pkg/hubble/filters/httpheader_test.go:219]:
```suggestion
			name: "header-key-mismatch",
```

File: pkg/hubble/filters/httpheader_test.go [hunk 24]
```diff
@@ -0,0 +1,337 @@
...
+						},
+					},
+				},
+			},
+
+			want: false,
+		},
+
+		{
+			name: "header_value_mismatch",
```
Comment (kaworu) [inline pkg/hubble/filters/httpheader_test.go:255]:
```suggestion
			name: "header-value-mismatch",
```

File: pkg/hubble/filters/httpheader_test.go [hunk 25]
```diff
@@ -0,0 +1,337 @@
...
+						},
+					},
+				},
+			},
+
+			want: false,
+		},
+
+		{
+			name: "header_key_and_value_mismatch",
```
Comment (kaworu) [inline pkg/hubble/filters/httpheader_test.go:292]:
```suggestion
			name: "header-key-and-value-mismatch",
```

File: pkg/hubble/filters/workload_test.go [hunk 26]
```diff
@@ -220,6 +220,7 @@ func TestWorkloadFilterInclude(t *testing.T) {
 			},
 			want: false,
 		},
+
```
Comment (kaworu) [inline pkg/hubble/filters/workload_test.go:223]:
```suggestion
```

File: pkg/hubble/filters/workload_test.go [hunk 27]
```diff
@@ -244,6 +245,7 @@ func TestWorkloadFilterInclude(t *testing.T) {
 			},
 			want: true,
 		},
+
```
Comment (kaworu) [inline pkg/hubble/filters/workload_test.go:248]:
```suggestion
```

**Review by kaworu — CHANGES_REQUESTED** [review]
Hi @rohithsai1904

1. The [Travis-CI failure](https://app.travis-ci.com/github/cilium/cilium/jobs/591464072) can be solved by running `make -C api/v1`
2. I left some suggestions, the one about the test arg struct should fix the compilation error. Then, you can run the following command to ensure that all new tests are passing and iterate from there:
    ```
   go test -v -run "TestHTTPHeaderFilter" ./pkg/hubble/filters/...
    ```
3. Optionally, it would be great to add an additional test case for multiple HTTP headers with the same name, e.g. a flow with both `Cache-control: no-cache` and `Cache-control: no-store` should match a `Cache-control: no-store` filter.
4. Finally, once the PR is passing CI and tests, please squash all commits into one and sign-off the resulting commit (see https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#developer-s-certificate-of-origin).

**Pull request closed without merging**